### PR TITLE
fix(3012 ): Refactoring builds triggers

### PR DESCRIPTION
**Pull request by y-oksaku**
## Context

The trigger processing is complex, making it difficult to modify some aspects of the trigger processing independently from others.
As a result, fixing bugs and unexpected behavior becomes challenging.

## Objective

This PR refactors the build trigger processing as follows: https://github.com/screwdriver-cd/screwdriver/issues/3012

- Separate each trigger into a class: OR trigger, AND trigger, RemoteTrigger, and RemoteJoin.
- Simplify these trigger processes.
- Move functions to a helper file.
- Add more tests for triggers.

Now, we can trigger a single remote join in the downstream (e.g. `requires: [ sd@1:a ]`).
However, some existing strange behaviors were found, and these have not been fixed. (Refer to the comments of https://github.com/screwdriver-cd/screwdriver/issues/3012)

## References

- Issue: https://github.com/screwdriver-cd/screwdriver/issues/3012
  - There are sub PRs that have been refactoring.

If something happened, please mention to us ([kumada626](https://github.com/kumada626), [yk634](https://github.com/yk634), [yakanechi](https://github.com/yakanechi), [y-oksaku](https://github.com/y-oksaku)).

## License
I confirm that this contribution is made under a BSD license and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

File: plugins/builds/triggers/or.js [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+ * @property {number} username
+ * @property {string} scmContext
+ */
+class OrTrigger extends OrBase {
+    /**
+     * Trigger the next jobs of the current job
+     * @param {EventModel} event
+     * @param {number} pipelineId
+     * @param {string} nextJobName
+     * @param {string} nextJobId
```
Comment (tkyi) [inline plugins/builds/triggers/or.js:26]:
```suggestion
     * @param {number} nextJobId
```

File: plugins/builds/triggers/or.js [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+ * @typedef {import('screwdriver-models').PipelineFactory} PipelineFactory
+ * @typedef {import('screwdriver-models/lib/build').BuildModel} BuildModel
+ * @typedef {import('screwdriver-models/lib/event').EventModel} EventModel
+ */
+/**
+ * @property {BuildFactory} buildFactory
+ * @property {JobFactory} jobFactory
+ * @property {PipelineFactory} pipelineFactory
+ * @property {BuildModel} currentBuild
+ * @property {number} username
```
Comment (tkyi) [inline plugins/builds/triggers/or.js:17]:
```suggestion
 * @property {string} username
```

File: plugins/builds/types/models.d.ts [hunk 3]
```diff
@@ -0,0 +1,476 @@
...
+    prParentJobId: undefined;
+    permutations: Array<Permutation>;
+    description: undefined;
+    pipelineId: number;
+    state: string;
+    stateChanger: undefined;
+    stateChangeTime: undefined;
+    stateChangeMessage: undefined;
+    archived: boolean;
+    templateId: undefined;
```
Comment (tkyi) [inline plugins/builds/types/models.d.ts:94]: are most of these meant to be `undefined`?

File: plugins/builds/types/index.d.ts [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+ * @property {number}               username        Triggered username
+ * @property {string}               scmContext      e.g., 'github:github.com'
+ * @property {EventModel}           event           Event to which the current build belongs
+ * @property {StageModel | null}    event           Stage to which the current build belongs
+ */
+export type ServerConfig = {
+    pipeline: PipelineModel;
+    job: JoinJob;
+    build: BuildModel;
+    username: number;
```
Comment (tkyi) [inline plugins/builds/types/index.d.ts:21]:
```suggestion
    username: string;
```

File: plugins/builds/types/index.d.ts [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+// -----------------------------------------------------------------------------
+/**
+ * The assigned arguments of hapi servers expose function
+ * @property {PipelineModel}        pipeline        Pipeline to which the current build belongs
+ * @property {JoinJob}              job             Job to which the current build belongs
+ * @property {BuildModel}           build           Build to which the current build belongs
+ * @property {number}               username        Triggered username
+ * @property {string}               scmContext      e.g., 'github:github.com'
+ * @property {EventModel}           event           Event to which the current build belongs
+ * @property {StageModel | null}    event           Stage to which the current build belongs
```
Comment (tkyi) [inline plugins/builds/types/index.d.ts:15]:
```suggestion
 * @property {StageModel | null}    stage           Stage to which the current build belongs
```

File: plugins/builds/types/models.d.ts (hunk 3 above)
Comment (y-oksaku) [reply plugins/builds/types/models.d.ts]: Sorry, these `.d.ts` files were already unnecessary :bow:

We used these files to understand models.
Thus some of them didn't have correct property's type.

I cleaned up these files and fixed JSDocs.

**Review by tkyi — APPROVED** [review]

**Pull request merged**